walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,107 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
+
+  /*
+   Criterion functional interface is used to define who's (mandatory or all attendees')
+   occupied hours should be taken into account in the findAvailableTimeRanges function.
+   */
+  @FunctionalInterface
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+  private final Criterion COMMON_CRITERION = x -> x == 0;
+  private final Criterion MANDATORY_CRITERION = x -> x <= 1;
+
+  private static final int MINUTES = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean haveCommonAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  /*
+  The algorithm skips all minutes that don't meet the criterion,
+  memorises the index of the first fitting minute in a row and goes until
+  it finds the new unfitting minute. Adds the range to the list if
+  it's not less than the target duration.
+   */
+  /**
```

Comment:
Consider merging these two comment sections into a single [Javadoc section](https://google.github.io/styleguide/javaguide.html#s7.1.1-javadoc-multi-line) (similarly to the section with the params), where the first paragraph describes the algorithm, followed by a blank line (starting with `*`), followed by the `@param`-s.